changes/changes_test.go

```diff
@@ -0,0 +1,550 @@
+	if diff := cmp.Diff(colItems, expectColItems); diff != "" {
+		t.Errorf("column items result mismatch. (-want +got):%s\n", diff)
+	}
+
+	expectSummary := StatsChangeSummaryFields{
+		Entries:   5,
+		Columns:   4,
+		TotalSize: 155,
+	}
+
+	if diff := cmp.Diff(summary, expectSummary); diff != "" {
+		t.Errorf("stats summary result mismatch. (-want +got):%s\n", diff)
+	}
+}
+
+func TestMaybeLoadStats(t *testing.T) {
+	ctx := context.Background()
+
+	workDir, err := ioutil.TempDir("", "qri_test_changes_service")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(workDir)
+
+	mr, err := repotest.NewTestRepo()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	svc := newTestService(t, mr, workDir)
+
+	ref := dsref.MustParse("peer/cities")
+	if _, err := mr.ResolveRef(ctx, &ref); err != nil {
+		t.Fatal(err)
+	}
+
+	ds, err := dsfs.LoadDataset(ctx, mr.Filesystem(), ref.Path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = base.OpenDataset(ctx, mr.Filesystem(), ds); err != nil {
+		t.Fatal(err)
+	}
+
+	if ds.Stats == nil {
+		t.Fatal("stats are nil")
+	}
+
+	ds.Stats = nil
+
+	svc.maybeLoadStats(ctx, ds)
+	if ds.Stats == nil {
+		t.Fatal("stats are nil")
+	}
+}
+
+func TestMatchColumns(t *testing.T) {
+	ctx := context.Background()
+
+	workDir, err := ioutil.TempDir("", "qri_test_changes_service")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(workDir)
+
+	mr, err := repotest.NewTestRepo()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	svc := newTestService(t, mr, workDir)
+
+	ref := dsref.MustParse("peer/cities")
+	if _, err := mr.ResolveRef(ctx, &ref); err != nil {
+		t.Fatal(err)
+	}
+
+	ds, err := dsfs.LoadDataset(ctx, mr.Filesystem(), ref.Path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = base.OpenDataset(ctx, mr.Filesystem(), ds); err != nil {
+		t.Fatal(err)
+	}
+
+	ds.Name = "cities"
+	leftDs := *ds
+
+	// alter body file
+	const alteredBodyData = `city,pop,avg_age,in_usa
+toronto,4000000,55.0,false
+new york,850000,44.0,false
+chicago,30000,440.4,false
+chatham,3500,650.25,false
+raleigh,25000,5000.65,false`
+
+	updateDataset(t, mr, ds, alteredBodyData)
+
+	var leftColItems tabular.Columns
+	_, err = svc.parseColumns(&leftColItems, &leftDs)
+	if err != nil {
+		t.Fatal(err)
+	}
+	leftStats, err := svc.parseStats(&leftDs)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var rightColItems tabular.Columns
+	_, err = svc.parseColumns(&rightColItems, ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+	rightStats, err := svc.parseStats(ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	report, err := svc.matchColumns(4, 4, leftColItems, rightColItems, leftStats, rightStats)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	reportJSON, _ := json.Marshal(report)
+
+	t.Logf("%s\n", reportJSON)
+
+	// alter body file - remove column
+	const alteredBodyDataColumns1 = `city,avg_age,in_usa
+toronto,55.0,false
+new york,44.0,false
+chicago,440.4,false
+chatham,650.25,false
+raleigh,5000.65,false`
+
+	ds.Name = "cities"
+
+	updateDataset(t, mr, ds, alteredBodyDataColumns1)
```

Comment:
Need some help with this, apparently this method doesn't properly update the `ds.Structure` while all else I needed seems fine. Functionally speaking things seems to be correct when I force the right type of problem for it to pass. @dustmop @b5 any hints on how to improve my `updateDataset` method?

Reply:
ah I think I understand. Are you expecting structure to change because you changed the body structure? currently we only infer a structure _if one doesn't exist_. So on the first commit a structure gets created, then needs to be explicitly overridden in future commits.

One easy solution: have your `updateDataset` function explicitly set ds.Structure to nil. `base.CreateDataset` will re-infer a new structure for you that'll match